addons/mail/models/mail_thread.py

```diff
@@ -1008,6 +1008,20 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
                     email_from, email_to, message_id, fallback_model, thread_id, custom_values, user_id)
                 return [route]
 
+        # 4. If catchall email is in email_to, we should bounce
+        if rcpt_tos_localparts:
+            # no route found for a matching reference (or reply), so parent is invalid
+            message_dict.pop('parent_id', None)
+
+            # check it does not directly contact catchall
+            if catchall_alias and email_to_localparts and any(email_localpart == catchall_alias for email_localpart in email_to_localparts):
```

Comment:
Actually I guess you could merge the condition as `message_dict.pop` is only relevant inside the condition anyway. Also note that `email_to_localparts` is always an iterable, and `any([])` is always `False` so `and email_to_local_parts` can be omitted from the condition.

```suggestion
        if rcpt_tos_localparts and catchall_alias and any(email_localpart == catchall_alias for email_localpart in email_to_localparts):
            # no route found for a matching reference (or reply), so parent is invalid
            message_dict.pop('parent_id', None)
```